services/lookupAddress.js

```diff
@@ -0,0 +1,100 @@
+// from https://github.com/ChainSafe/web3.js/issues/2683#issuecomment-547348416
+
+import namehash from 'eth-ens-namehash'
+import { BigNumber, utils } from 'ethers'
+import ABI from 'web3-eth-ens/lib/resources/ABI/Resolver'
+import uniq from 'lodash/uniq'
+
+export const createBatchRequestCallback = (resolve, reject) => (error, data) => {
+  if (error) {
+    reject(error)
+  } else {
+    resolve(data)
+  }
+}
+
+const CACHE = {}
+
+const createFetchNodeAddresses = (registryContract, batch) => async (address) => {
+  const addressLower = address.toLowerCase()
+
+  const node = addressLower.substr(2) + '.addr.reverse'
+  const nodeHash = namehash.hash(node)
+  let nodeAddress = null
+
+  if (!CACHE[addressLower]) {
+    try {
+      nodeAddress = await new Promise((resolve, reject) => {
+        const callback = createBatchRequestCallback(resolve, reject)
+        const requestData = registryContract.methods.resolver(nodeHash).call.request(callback)
+        batch.add(requestData)
+      })
+
+      if (+nodeAddress === 0) nodeAddress = null
+    } catch (error) {
+      console.error(`Error resolve ens for "${address}"`, error.message)
+      // do nothing
+    }
+  }
+
+  return {
+    addressLower,
+    address,
+    nodeHash,
+    nodeAddress
+  }
+}
+
+const createFetchEnsNames = (web3, batch, results) => async (data) => {
+  const { address, addressLower, nodeHash, nodeAddress } = data
+  if (!nodeAddress) return results
+
+  if (CACHE[addressLower]) {
+    results[address] = CACHE[addressLower]
+    return results
+  }
+
+  const nodeContract = new web3.eth.Contract(ABI, nodeAddress)
+
+  try {
+    const ensName = await new Promise((resolve, reject) => {
+      const callback = createBatchRequestCallback(resolve, reject)
+      const requestData = nodeContract.methods.name(nodeHash).call.request(callback)
+      batch.add(requestData)
+    })
+
+    const isZeroAddress =
+      ensName.trim().length && utils.isAddress(ensName) && BigNumber.from(ensName).isZero()
+
+    if (isZeroAddress) return results
+
+    CACHE[addressLower] = ensName
+    results[address] = ensName
+
+    return results
+  } catch (error) {
+    console.error(`Error lookupAddress ens for "${address}"`, error.message)
+    return results
+  }
+}
+
+export const lookupAddresses = async (addressList, web3) => {
+  web3.eth.ens._lastSyncCheck = Date.now() // - need for test in fork
+
+  const registryContract = await web3.eth.ens.registry.contract
+
+  const fetchNodeAddressesBatch = new web3.BatchRequest()
```

Comment:
Do we need to limit the number of batch requests? A large number of requests may not pass

Reply:
Proposal 13
fetchVotedEvents 34
fetchVotedEvents uniq 32

Proposal 12
fetchVotedEvents 44
fetchVotedEvents uniq 41

Proposal 11
fetchVotedEvents 29
fetchVotedEvents uniq 26

Proposal 10
fetchVotedEvents 52
fetchVotedEvents uniq 44

Proposal 9
fetchVotedEvents 16
fetchVotedEvents uniq 14

Proposal 8
fetchVotedEvents 11
fetchVotedEvents uniq 11

Proposal 7
fetchVotedEvents 46
fetchVotedEvents uniq 40

Proposal 6
fetchVotedEvents 29
fetchVotedEvents uniq 28

Proposal 5
fetchVotedEvents 52
fetchVotedEvents uniq 50

Proposal 4
fetchVotedEvents 69
fetchVotedEvents uniq 61

Proposal 3
fetchVotedEvents 17
fetchVotedEvents uniq 16

Proposal 2
fetchVotedEvents 31
fetchVotedEvents uniq 30

Proposal 1
fetchVotedEvents 118
fetchVotedEvents uniq 111

Reply:
done for batch and add TODO